exec/java-exec/src/main/java/org/apache/drill/exec/record/metadata/AbstractColumnMetadata.java

```diff
@@ -0,0 +1,206 @@
+import org.apache.drill.exec.record.MaterializedField;
+
+/**
+ * Abstract definition of column metadata. Allows applications to create
+ * specialized forms of a column metadata object by extending from this
+ * abstract class.
+ * <p>
+ * Note that, by design, primitive columns do not have a link to their
+ * tuple parent, or their index within that parent. This allows the same
+ * metadata to be shared between two views of a tuple, perhaps physical
+ * and projected views. This restriction does not apply to map columns,
+ * since maps (and the row itself) will, by definition, differ between
+ * the two views.
+ */
+
+public abstract class AbstractColumnMetadata implements ColumnMetadata {
+
+  // Capture the key schema information. We cannot use the MaterializedField
+  // or MajorType because then encode child information that we encode here
+  // as a child schema. Keeping the two in sync is nearly impossible.
+
+  protected final String name;
+  protected final MinorType type;
+  protected final DataMode mode;
+  protected final int precision;
+  protected final int scale;
+  protected boolean projected = true;
+
+  /**
+   * Predicted number of elements per array entry. Default is
+   * taken from the often hard-coded value of 10.
+   */
+
+  protected int expectedElementCount = 1;
+
+  public AbstractColumnMetadata(MaterializedField schema) {
+    name = schema.getName();
+    MajorType majorType = schema.getType();
+    type = majorType.getMinorType();
+    mode = majorType.getMode();
+    precision = majorType.getPrecision();
+    scale = majorType.getScale();
+    if (isArray()) {
+      expectedElementCount = DEFAULT_ARRAY_SIZE;
+    }
+  }
+
+  public AbstractColumnMetadata(String name, MinorType type, DataMode mode) {
+    this.name = name;
+    this.type = type;
+    this.mode = mode;
+    precision = 0;
+    scale = 0;
+    if (isArray()) {
+      expectedElementCount = DEFAULT_ARRAY_SIZE;
+    }
+  }
+
+  public AbstractColumnMetadata(AbstractColumnMetadata from) {
+    name = from.name;
+    type = from.type;
+    mode = from.mode;
+    precision = from.precision;
+    scale = from.scale;
+    expectedElementCount = from.expectedElementCount;
+  }
+
+  protected void bind(TupleSchema parentTuple) { }
+
+  @Override
+  public String name() { return name; }
+
+  @Override
+  public MinorType type() { return type; }
+
+  @Override
+  public MajorType majorType() {
+    return MajorType.newBuilder()
+        .setMinorType(type())
+        .setMode(mode())
+        .build();
+  }
+
+  @Override
+  public DataMode mode() { return mode; }
+
+  @Override
+  public boolean isNullable() { return mode() == DataMode.OPTIONAL; }
+
+  @Override
+  public boolean isArray() { return mode() == DataMode.REPEATED; }
+
+  @Override
+  public int dimensions() { return isArray() ? 1 : 0; }
+
+  @Override
+  public boolean isMap() { return false; }
+
+  @Override
+  public boolean isVariant() { return false; }
```

Comment:
I think there is a write-up somewhere about this. "Variant" is an old term from Visual Basic for a tagged union (which is what our "Union" type really is.) It is used here because both a (non-repeated) List and a Union are both unions, which is confusing, so I used a separate term. Also, the "Union" term is confusing since there is also a UNION operator in SQL. `See VariantMetadata.java`.